### ISSUE TITLE
torch.compile/triton holding GIL during compilation and CompiledKernel call results in deadlocks during distributed training

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#109074]
### 🐛 Describe the bug

We are seeing our training processed get deadlocked as follows. There is a NCCL error causing the allreduce in the backward pass to get stuck. When this occurs, typically the GPU locks up until a [ncclCommAbort](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommabort) is called and even `cudaLaunchKernel` gets stuck on the host side.  As a result, the main thread is stuck as follows:

```
Thread 432919 (idle): "MainThread"
    0x155555277197 (libc.so.6)
    pthread_cond_wait (libc.so.6)
    __gnu_cxx::__nothrow_wait_cv::wait (compatibility-condvar.cc:100)
    torch::autograd::ReadyQueue::pop (libtorch_cpu.so)
    torch::autograd::Engine::thread_main (libtorch_cpu.so)
    torch::autograd::Engine::execute_with_graph_task (libtorch_cpu.so)
    torch::autograd::python::PythonEngine::execute_with_graph_task (libtorch_python.so)
    torch::autograd::Engine::execute (libtorch_cpu.so)
    torch::autograd::python::PythonEngine::execute (libtorch_python.so)
    THPEngine_run_backward (libtorch_python.so)
    backward (torch/autograd/__init__.py:200)
    backward (torch/_tensor.py:503)
  ```
  
  Simultaneously, triton is stuck due to the above NCCL issue while trying to call a compiled kernel. The bad part here though is that triton is holding GIL (see active+gil in the trace below) and is stuck. As a result, the entire Python process freezes up and no other python threads are able to execute:

```
Thread 493983 (active+gil): "Dummy-20"
    sched_yield (libc.so.6)
    0x15550de3cbdb (libcuda.so.535.54.03)
    0x15550e10adb5 (libcuda.so.535.54.03)
    0x15550e10b32a (libcuda.so.535.54.03)
    0x15550df18454 (libcuda.so.535.54.03)
    0x15550ddf7836 (libcuda.so.535.54.03)
    0x15550ddf7f3d (libcuda.so.535.54.03)
    0x15550ddfaa27 (libcuda.so.535.54.03)
    0x15550dfe9b3e (libcuda.so.535.54.03)
    launch (triton_.so)
    launcher (<string>:6)
    run (torch/_inductor/triton_ops/autotune.py:190)
    call (coksmpcybzxz4bxfvl56kcd6k5czlqpovjyjwg4yry6do7p7hcim.py:871)
    run (torch/_inductor/compile_fx.py:248)
    _fn (torch/_dynamo/eval_frame.py:209)
    call_func_with_args (torch/_functorch/aot_autograd.py:1249)
    call_compiled_backward (torch/_functorch/aot_autograd.py:2341)
    backward (torch/_functorch/aot_autograd.py:2365)
```

Another example is that triton also get stuck during compilation as well due to a `torch.cuda.synchronize` getting stuck as NCCL kernels are stuck. Even in this case, triton is stuck while holding the GIL and results in the entire Python process to lock up:

```
Thread 1957540 (active+gil): "Dummy-20"
    synchronize (torch/cuda/__init__.py:688)
    _precompile_config (torch/_inductor/triton_ops/autotune.py:94)
    <listcomp> (torch/_inductor/triton_ops/autotune.py:70)
    precompile (torch/_inductor/triton_ops/autotune.py:69)
    _load_kernel (torch/_inductor/codecache.py:554)
    result (torch/_inductor/codecache.py:574)
    wait (torch/_inductor/codecache.py:715)
    <module> (cpb2lazn3hvywa5maow5ngritvyt6pwlcq6ela3fl5rbblyj636x.py:1578)
    load (torch/_inductor/codecache.py:528)
    compile_to_module (torch/_inductor/graph.py:575)
    time_wrapper (torch/_dynamo/utils.py:163)
    compile_to_fn (torch/_inductor/graph.py:586)
    compile_fx_inner (torch/_inductor/compile_fx.py:177)
    inner (contextlib.py:75)
    inner (torch/_inductor/debug.py:239)
    debug_wrapper (torch/_dynamo/debug_utils.py:595)
    bw_compiler (torch/_inductor/compile_fx.py:441)
    time_wrapper (torch/_dynamo/utils.py:163)
    _fn (torch/_dynamo/eval_frame.py:209)
    _wrapped_bw_compiler (torch/_dynamo/backends/common.py:38)
    call_compiled_backward (torch/_functorch/aot_autograd.py:2336)
    backward (torch/_functorch/aot_autograd.py:2365)
    apply (torch/autograd/function.py:274)
```

We have some custom logic in our application to recover from NCCL errors by calling `_abort` on the ProcessGroup here: https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/init.cpp#L2226 (We set `NCCL_ASYNC_ERROR_HANDLING=0` and manage errors ourselves). This logic is usually able to recover from the NCCL errors. However, now since triton is holding GIL and is stuck, the entire Python process is frozen and none our custom logic in python can run and recover from the situation.

I'm wondering if it is possible to release GIL before entering C/C++ land (similar to other PyTorch operations). This would resolve the issue and allow other python threads in the application to recover from the situation.

### Versions

PyTorch 2.0.1

**Comment by pritamdamania87** [comment]
@eellison Based on your previous https://github.com/pytorch/pytorch/issues/108586#issuecomment-1711862033, I'm wondering if [this](https://github.com/openai/triton/compare/main...eellison:triton:multi_thread?expand=1) work would resolve the above issue as well?

**Comment by pritamdamania87** [comment]
Looks like triton GIL on launch was resolved here: https://github.com/openai/triton/pull/2185, but the issue with torchinductor calling `torch.cuda.synchronize` still exists: https://github.com/pytorch/pytorch/blob/293d3b89d8bf8f966fe06d25ac1089a3739e20e3/torch/_inductor/triton_heuristics.py#L200

**Comment by ezyang** [comment]
eellison's PR would get rid of the GIL for Triton compilation, but it's not clear this gets rid of all the places we get rid of the GIL. (Does torch.cuda.synchronize hold the GIL? Ed: I feel like it shouldn't.)

**Comment by eellison** [comment]
Actually, triton compilation already doesn't hold gil (although it isn't parallelized).

**Comment by pritamdamania87** [comment]
> Does torch.cuda.synchronize hold the GIL? Ed: I feel like it shouldn't.

It does seem to hold the GIL (see the traceback above). I looked through the code as well and didn't spot any place where we are releasing the GIL: https://github.com/pytorch/pytorch/blob/main/torch/csrc/cuda/Module.cpp#L444. Unless I'm missing something, it seems like most of `torch._C._cuda` calls don't release GIL.

**Comment by pritamdamania87** [comment]
@ezyang I have a fix in https://github.com/pytorch/pytorch/pull/109159

**Comment by pritamdamania87** [comment]
sigh, another place where we don't release GIL:

```
Thread 1642008 (active+gil): "Dummy-22"
    sched_yield (libc.so.6)
    0x15550de3cbdb (libcuda.so.535.54.03)
    0x15550e10adb5 (libcuda.so.535.54.03)
    0x15550e10b32a (libcuda.so.535.54.03)
    0x15550df18454 (libcuda.so.535.54.03)
    0x15550ddf7836 (libcuda.so.535.54.03)
    0x15550ddf7f3d (libcuda.so.535.54.03)
    0x15550ddfaa27 (libcuda.so.535.54.03)
    0x15550dfe9b3e (libcuda.so.535.54.03)
    0x15551181490c (libcudart.so.12.0.146)
    cudaLaunchKernel (libcudart.so.12.0.146)
    at::native::gpu_kernel_impl<at::native::FillFunctor<float> > (libtorch_cuda.so)
    at::native::gpu_kernel<at::native::FillFunctor<float> > (libtorch_cuda.so)
    at::native::fill_kernel_cuda(at::TensorIterator&, c10::Scalar const&)::{lambda()#1}::operator() const (libtorch_cuda.so)
    at::native::fill_kernel_cuda (libtorch_cuda.so)
    at::native::fill_out (libtorch_cpu.so)
    at::(anonymous namespace)::(anonymous namespace)::wrapper_CUDA_Scalar_fill_ (libtorch_cuda.so)
    at::_ops::fill__Scalar::call (libtorch_cpu.so)
    at::native::zero_ (libtorch_cpu.so)
    at::(anonymous namespace)::(anonymous namespace)::wrapper_CUDA__zero_ (libtorch_cuda.so)
    torch::ADInplaceOrView::(anonymous namespace)::zero_ (libtorch_cpu.so)
    torch::autograd::VariableType::(anonymous namespace)::zero_ (libtorch_cpu.so)
    at::_ops::zero_::call (libtorch_cpu.so)
    at::native::zeros_symint (libtorch_cpu.so)
    c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>), &at::(anonymous namespace)::(anonymous namespace)::wrapper_CompositeExplicitAutograd__zeros(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> > >, at::Tensor(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)>::call (libtorch_cpu.so)
    at::_ops::zeros::redispatch (libtorch_cpu.so)
    c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoFunctor_<c10::CompileTimeFunctionPointer<at::Tensor(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>), &at::(anonymous namespace)::zeros(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)>, at::Tensor, c10::guts::typelist::typelist<c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool> > >, at::Tensor(c10::ArrayRef<c10::SymInt>, c10::optional<c10::ScalarType>, c10::optional<c10::Layout>, c10::optional<c10::Device>, c10::optional<bool>)>::call (libtorch_cpu.so)
    at::_ops::zeros::call (libtorch_cpu.so)
    torch::autograd::VariableInfo::zeros (libtorch_cpu.so)
    torch::autograd::PyNode::apply (libtorch_python.so)
    torch::autograd::Node::operator() (libtorch_cpu.so)
    torch::autograd::Engine::evaluate_function (libtorch_cpu.so)
    torch::autograd::Engine::thread_main (libtorch_cpu.so)
    torch::autograd::Engine::thread_init (libtorch_cpu.so)
    torch::autograd::python::PythonEngine::thread_init (libtorch_python.so)
    execute_native_thread_routine (thread.cc:106)
    0x15555527ab43 (libc.so.6)
    0x15555530ca00 (libc.so.6)
```

@Chillee I'm assuming the above is because AOT Autograd is using PyNode ?

**Comment by pritamdamania87** [comment]
@ezyang @wconstab Thanks a lot for helping out with this issue and https://github.com/pytorch/pytorch/pull/109159. Although, I suspect this is probably not the last time we will encounter such an issue. We have already hit this issue in 3 separate instances, fixed by these PRs:

https://github.com/pytorch/pytorch/pull/104103
https://github.com/pytorch/pytorch/pull/109454
https://github.com/pytorch/pytorch/pull/109159

I was wondering if it would be possible to put in a proper long term fix for this issue? One extreme idea I had was to put an assert like `TORCH_ASSERT_NO_GIL_WITHOUT_PYTHON_DEP` before this line : https://github.com/pytorch/pytorch/blob/main/c10/cuda/CUDAException.h#L29 to validate we release GIL always and fix everything that breaks. We could have a test time assert to avoid running this check always and rely on test coverage to catch any issues.

**Comment by ezyang** [comment]
Well, `c10/util/DeadlockDetection.h` is designed to help you banish calls to GIL in APIs. The extreme idea is probably a bad idea because this test isn't all that fast, so you would need to have a DEBUG only variant of the assert (and get CI coverage on it), but I am definitely amenable to adding this to some critical APIs that you definitely don't want called with GIL.

**Comment by pritamdamania87** [comment]
> but I am definitely amenable to adding this to some critical APIs that you definitely don't want called with GIL.

The thing is that we don't really know which API these issues might come from. We are just using high-level torch APIs like `.backward()`, `torch.compile` etc. The issues mostly come from these higher-level APIs calling lower-level APIs which don't release GIL properly. 

> so you would need to have a DEBUG only variant of the assert (and get CI coverage on it)

I was mostly thinking of something like this so that the check only runs during CI and doesn't impact performance for users.

**Comment by wconstab** [comment]
it seems like if we're gonna land changes to remove GIL form random APIs whenever they cause deadlocks, then we're better off having the proposed CI test that lets us capture them all and hold the fort.  How expensive is it going to be really though?  Can we do a run and see?

**Comment by voznesenskym** [comment]
@pritamdamania87 please confirm if this is still happening

**Comment by pritamdamania87** [comment]
@voznesenskym We haven't seen GIL deadlocks recently. However, as I've mentioned in https://github.com/pytorch/pytorch/issues/109074#issuecomment-1734232417, the current state of affairs is very fragile and things can easily break again the future.

We need a proper long term fix for this issue.

**Comment by ngimel** [comment]
I'd like to +1 request for gil tests, it's all too easy to forget to drop the gil somewhere. Note #68739 discussed potential ways but it went nowhere. I'm not sure how expensive running the tests with debug build checking for gil in CUDAException would be, but if that's too expensive, we could maybe LD_PRELOAD `cudaLaunchKernel/cuLaunchKernel` for tests to blow up if gil is held? That would have caught https://github.com/pytorch/pytorch/pull/104103
and https://github.com/pytorch/pytorch/pull/109454 (the latter if we have autograd tests for multiple output autograd functions)

**Comment by gchanan** [comment]
explanation for removing high priority label:
- immediate issue has been addressed
- we'll prioritize reported fixing cases where we are unexpectedly / incorrectly holding the GIL
- better testing is always good but we don't have a prioritized workstream to identify these cases right now (if someone out there wants to help with this, let us know)
- [perhaps somewhat controversial] it may be higher leverage for PT to focus on building better detection/recovery mechanisms outside of process/python due to susceptibility to GIL issues.

**Comment by pritamdamania87** [comment]
We ran into another very painful GIL deadlock due to gil not being released here: https://github.com/pytorch/pytorch/blob/main/torch/csrc/cuda/Event.cpp#L125C28-L125C34.

@gchanan @ezyang @wconstab Curious if there are any concrete plans to have 100% coverage in terms of releasing GIL appropriately.

**Comment by wconstab** [comment]
No concrete plans, it is a good idea though. Any thoughts how to set up a test infra for catching them all efficiently so we could burn them down?

**Comment by pritamdamania87** [comment]
I did have a suggestion above in https://github.com/pytorch/pytorch/issues/109074#issuecomment-1734232417, where we add a debug test-time check to ensure GIL is always released as part of `C10_CUDA_CHECK`.

**Comment by aorenste** [comment]
Issue Scrubbing: The original issue as described was fixed but the longer-term question is whether we need some more comprehensive testing/verification and what's the proper way to do that. Note that with python 3.15 "free-threaded" python is available - maybe the fix is to just ensure we can run properly with that mode?
